modules/develop/pages/consume-data/consumer-offsets.adoc

```diff
@@ -4,20 +4,127 @@
+
+[source,java]
+----
+consumer.subscribe(Arrays.asList("foo", "bar"));
+while (true) {
+    ConsumerRecords<String, String> records = consumer.poll(100);
+    for (ConsumerRecord<String, String> record : records) {
+        // process records here ...
+        
+        // ... and at the appropriate point, call commit (not after every message)
+        consumer.commitSync();
+    }
+}
+----
+
+==== Asynchronous commit
+
+The advantage of asynchronous commits is lower latency, because the consumer does not pause to wait for the commit response. However, there is no automatic retry of the commit (`commitAsync`) if it fails. There is also increased coding complexity (due to the asynchronous callbacks). The following example shows an asynchronous commit in which the consumer will not block. Instead, the commit call registers a callback, which is executed once the commit returns:
+
+[source,java]
+----
+void callback() {
+    // executed when the commit returns
+}
+
+consumer.subscribe(Arrays.asList("foo", "bar"));
+while (true) {
+    ConsumerRecords<String, String> records = consumer.poll(100);
+    for (ConsumerRecord<String, String> record : records) {
+        // process records here ...
+        
+        // ... and at the appropriate point, call commit
+        consumer.commitAsync(callback);
+    }
+}
+----
+
+=== External offset management
+
+The external offset management strategy allows consumers to manage offsets independently of Redpanda. In this approach:
+
+- Consumers bypass the consumer group API and directly assign partitions instead of subscribing to a topic.
+- Offsets are not committed to Redpanda, but are instead stored in an external storage system.
+
+To implement an external offset management strategy:
+
+. Set `enable.auto.commit` to `false`.
+. Use `assign(Collection<TopicPartition>)` to assign partitions. 
+. Use the offset provided with each ConsumerRecord to save your position.
+. Upon restart, use `seek(TopicPartition, long)` to restore the position of the consumer.
+
+=== Hybrid offset management
+
+The hybrid offset management strategy allows consumers to handle their own consumer rebalancing while still leveraging Redpanda's offset commit functionality. In this approach:
+
+- Consumers bypass the consumer group API and directly assign partitions instead of subscribing to a topic.
+- Offsets are committed to Redpanda.
+
+== Offset commit best practices
+
+Follow these best practices to optimize offset commits.
+
+=== Avoid over-committing
+
+The purpose of a commit is to save consumer progress. More frequent commits reduce the amount of data to re-read after an application restart, as the commit interval directly affects the Recovery Point Objective (RPO). Because a lower RPO is desirable, application designers may believe that committing frequently is a good design choice.
```

Comment:
```suggestion
The purpose of a commit is to save consumer progress. More frequent commits reduce the amount of data to re-read after an application restart, as the commit interval directly affects the recovery point objective (RPO). Because a lower RPO is desirable, application designers may believe that committing frequently is a good design choice.
```